fri/src/matrix_reducer.rs

```diff
@@ -0,0 +1,188 @@
+use alloc::vec;
+use alloc::vec::Vec;
+use itertools::Itertools;
+use p3_field::{AbstractField, ExtensionField, Field, PackedField};
+use p3_matrix::MatrixRows;
+use p3_maybe_rayon::{IndexedParallelIterator, MaybeIntoParIter, ParallelIterator};
+use p3_util::indices_arr;
+use tracing::{info_span, instrument};
+
+// seems to be the sweet spot, could be tweaked based on benches
+const BATCH_SIZE: usize = 8;
+
+/// Optimized matrix reducer. Only works for binomial extension fields, or extension fields
+/// in which multiplication and addition with a base field element are simple elementwise operations.
```

Comment:
I started off with this an associated type on Extension, so BinomialExtension could have this Reducer, another extension (???) could have another, but then `p3-field` depended on `p3-matrix`, etc.

Reply:
I hope there's no arcane field extension where this isn't true

Reply:
Hm yeah isn't scalar (base field) multiplication always elementwise?

Reply:
it is for extensions over irreducible polys, I don't know anything about other types of extensions but I assume we don't need them

Reply:
Scalar multiplication will always be elementwise as extensions are vector spaces over the base field.

If you quotient by something that isn't irreducible the vector space structure is still there you just find that the resulting space is a ring with zero divisors and not a field.